zh/tidb-database-development-specification/unstable-result-set.md

```diff
@@ -0,0 +1,189 @@
+| 2018_CS_03 | PatrickStar  | Physics                 |       6.0 |
++------------+--------------+-------------------------+-----------+
+36 rows in set (0.01 sec)
+
+```
+
+当遇到相同的order by值时，排序结果不稳定。为减少随机性，应当尽可能保持order by值的唯一性。不能保证唯一的继续加，保证order by 的字段组合是唯一时，结果才能唯一。
+
+## 3. 由于 group_concat() 中没有使用 order by 导致结果集不稳定
+
+结果集不稳定是因为 TiDB 是并行地从存储层读取数据，所以 group_concat() 在不加 order by 的情况下得到的结果集展现顺序容易被感知到不稳定。
+
+group_concat() 要获取到按顺序输出的结果集，需要把用于排序的字段添加到 order by 子句中，这样才符合 SQL 的语义。在下面的案例中，使用 group_concat() 不加 order by 的情况下拼接 customer_id ，造成结果集不稳定：
+
+1）不加 order by
+
+第一次查询：
+
+```sql
+mysql>  select GROUP_CONCAT( customer_id SEPARATOR ',' ) FROM customer where customer_id like '200002%';
++-------------------------------------------------------------------------+
+| GROUP_CONCAT(customer_id  SEPARATOR ',')                                |
++-------------------------------------------------------------------------+
+| 20000200992,20000200993,20000200994,20000200995,20000200996,20000200... |
++-------------------------------------------------------------------------+
+
+```
+
+第二次查询：
+
+```sql
+mysql>  select GROUP_CONCAT( customer_id SEPARATOR ',' ) FROM customer where customer_id like '200002%';
++-------------------------------------------------------------------------+
+| GROUP_CONCAT(customer_id  SEPARATOR ',')                                |
++-------------------------------------------------------------------------+
+| 20000203040,20000203041,20000203042,20000203043,20000203044,20000203... |
++-------------------------------------------------------------------------+
+
+```
+
+2）加 order by
+
+第一次查询：
+
+```sql
+mysql>  select GROUP_CONCAT( customer_id order by customer_id SEPARATOR ',' ) FROM customer where customer_id like '200002%';
++-------------------------------------------------------------------------+
+| GROUP_CONCAT(customer_id  SEPARATOR ',')                                |
++-------------------------------------------------------------------------+
+| 20000200000,20000200001,20000200002,20000200003,20000200004,20000200... |
++-------------------------------------------------------------------------+
+
+```
+
+第二次查询：
+
+```sql
+mysql>  select GROUP_CONCAT( customer_id order by customer_id SEPARATOR ',' ) FROM customer where customer_id like '200002%';
++-------------------------------------------------------------------------+
+| GROUP_CONCAT(customer_id  SEPARATOR ',')                                |
++-------------------------------------------------------------------------+
+| 20000200000,20000200001,20000200002,20000200003,20000200004,20000200... |
++-------------------------------------------------------------------------+
+
+```
+
+4\. select \* from t limit 10 的结果依赖于数据插入顺序
+
+对简单查询如 \`select \* from t limit 10\`，其返回结果和数据插入顺序有关。如果数据重新导入一遍，顺序变了，那select \* from t limit 10 结果哪些是对的，语义不清楚，结果会不固定。
+
+5\. rand函数
+
+返回一个随机浮点值。其本身返回的结果是随机的。
+
+6、UDF函数
+
+用户自定义函数，TiDB 无法保证唯一。
+
+7、正则表达式函数
+
+不同的库实现方式不同，但都符合语义。如果换了库则会出现结果不一致的情况。

```

Comment:
`rand` should be random, `UDF` not supported yet and `regex` or sth it should be stable.

Reply:
deleted